RIOT-OS/RIOT
# cpu/rpx0xx: initial PIO support

### Contribution description



This adds a new `periph` interface for Programmable IO `(PIO)` and implements PIO
for the `rpi-pico`. To test basic functionality, there is a test application:
`tests/periph_pio/`. To test further, there is also a `PIO` implementation for
RIOTs `I2C` interface which should allow you to run tests for `I2C` device drivers.

We need the [pioasm](https://github.com/raspberrypi/pico-sdk/tree/master/tools/pioasm) from the `pico-sdk`, to assemble PIO programs. <del>I compiled the assembler and copied the binary to `/usr/local/bin/pioasm`</del>. It is downloaded from github and compiled during the build process if needed. The bash script should work with all PIO programs from [pico-examples](https://github.com/raspberrypi/pico-examples/tree/master/pio).

### Testing procedure



```
$ cd tests/driver_at24cxxx
$ CFLAGS+="-DPIO_I2C_CONFIG=\{.sda=2,.scl=3,.irq=0\}" BOARD=rpi-pico make flash term
```

```
2021-12-18 21:03:32,720 # main(): This is RIOT! (Version: 2021.10-devel-1049-gd35a26-rp2040_PIO)
2021-12-18 21:03:32,723 # Starting tests for module at24cxxx
2021-12-18 21:03:32,725 # EEPROM size: 32768 byte
2021-12-18 21:03:32,727 # Page size  : 64 byte
2021-12-18 21:03:32,729 # [SUCCESS] at24cxxx_init
2021-12-18 21:03:32,732 # [SUCCESS] at24cxxx_write_byte
2021-12-18 21:03:32,735 # [SUCCESS] at24cxxx_read_byte
2021-12-18 21:03:32,738 # [SUCCESS] write_byte/read_byte
2021-12-18 21:03:32,744 # [SUCCESS] at24cxxx_write
2021-12-18 21:03:32,747 # [SUCCESS] at24cxxx_read
2021-12-18 21:03:32,749 # [SUCCESS] write/read
2021-12-18 21:03:32,755 # [SUCCESS] at24cxxx_set
2021-12-18 21:03:32,760 # [SUCCESS] set/read
2021-12-18 21:03:32,763 # Finished tests for module at24cxxx
```

See also logic analyzer output:  
[at24cxxx.zip](https://github.com/RIOT-OS/RIOT/files/7740722/at24cxxx.zip)

### Issues/PRs references



#15822
